engines/cryomni3d/versailles/engine.h

```diff
@@ -0,0 +1,632 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#ifndef CRYOMNI3D_VERSAILLES_ENGINE_H
+#define CRYOMNI3D_VERSAILLES_ENGINE_H
+
+#include "common/events.h"
+#include "common/random.h"
+#include "common/array.h"
+#include "common/hashmap.h"
+#include "common/str.h"
+
+#include "cryomni3d/cryomni3d.h"
+#include "cryomni3d/omni3d.h"
+#include "cryomni3d/sprites.h"
+#include "cryomni3d/wam_parser.h"
+
+#include "cryomni3d/versailles/documentation.h"
+#include "cryomni3d/versailles/toolbar.h"
+#include "cryomni3d/versailles/dialogs_manager.h"
+
+namespace Graphics {
+class ManagedSurface;
+struct Surface;
+}
+
+namespace CryOmni3D {
+struct FixedImageConfiguration;
+class ZonFixedImage;
+}
+
+namespace CryOmni3D {
+namespace Versailles {
+struct PlaceStateActionKey {
+	unsigned int placeId;
+	unsigned int placeState;
+	unsigned int actionId;
+	PlaceStateActionKey(unsigned int placeId_, unsigned int placeState_, unsigned int actionId_) :
+		placeId(placeId_), placeState(placeState_), actionId(actionId_) {}
+
+	bool operator==(const PlaceStateActionKey &other) const {
+		return other.placeId == placeId && other.placeState == placeState && other.actionId == actionId;
+	}
+};
+
+struct PlaceActionKey {
+	unsigned int placeId;
+	unsigned int actionId;
+	PlaceActionKey(unsigned int placeId_, unsigned int actionId_) :
+		placeId(placeId_), actionId(actionId_) {}
+
+	bool operator==(const PlaceActionKey &other) const {
+		return other.placeId == placeId && other.actionId == actionId;
+	}
+};
+}
+}
+
+namespace Common {
+template<>
+struct Hash<CryOmni3D::Versailles::PlaceStateActionKey> {
+	uint operator()(const CryOmni3D::Versailles::PlaceStateActionKey &k) const {
+		// placeState shouldn't be greater than 8 and placeId shouldn't be greater than 100
+		// originalActionId shouldn't be greater than 65536
+		return (k.placeId << 24 | k.placeState << 16) ^ k.actionId;
+	}
+};
+template<>
+struct Hash<CryOmni3D::Versailles::PlaceActionKey> {
+	uint operator()(const CryOmni3D::Versailles::PlaceActionKey &k) const {
+		// placeId shouldn't be greater than 100
+		// originalActionId shouldn't be greater than 65536
+		return (k.placeId << 16) ^ k.actionId;
+	}
+};
+}
+
+namespace CryOmni3D {
+namespace Versailles {
+
+class CryOmni3DEngine_Versailles;
+
+enum AbortCommand {
+	AbortNoAbort = 0,
+	AbortQuit = 1,
+	AbortLoadGame = 2,
+	AbortNewGame = 3,
+	AbortNextLevel = 5,
+	AbortFinished = 6,
+	AbortGameOver = 7
+};
+
+struct GameVariables {
+	enum Var {
+		kCollectScore = 0,             // 0
```

Comment:
I recommend renaming these to kVar... to vaoid confusion with other enums.

Reply:
I put them in GameVariables structure to avoid clash. These enum members can only be accessed using GameVariables::kStuff.